[Merged by Bors] - feat(linear_algebra/matrix): define equivalences for reindexing matrices with equivalent types

ocfnash opened the pull request:
---
I have started some work to define and prove basic facts about the classical Lie algebras and I need `matrix.reindex_lie_equiv`.

Commit:
feat(linear_algebra/matrix): define equivalences for reindexing matrices with equivalent types

kim-em commented on src/linear_algebra/matrix.lean:
```diff
@@ -413,6 +413,56 @@ by rw [@linear_equiv.findim_eq R (matrix m n R) _ _ _ _ _ _ (linear_equiv.uncurr
+
+/-- The natural map that reindexes a matrix's rows and columns with equivalent types is a linear
+equivalence. -/
+def reindex_linear_equiv [semiring R] (eₘ : m ≃ m') (eₙ : n ≃ n') :
+  matrix m n R ≃ₗ[R] matrix m' n' R :=
+{ map_add'  := λ M N, rfl,
+  map_smul' := λ M N, rfl,
+..(reindex eₘ eₙ)}
+
```
I think you need a `@[simp]` lemma (also for the algebra and Lie algebra versions) that says what happens when you apply this equivalence to an argument.

ocfnash replied:
Sure, now done!

Commit:
Add missing simp lemmas

Commit:
Fix simp failure

After introducing matrix.reindex_linear_equiv_apply we must take
slightly more care to get matrix.reindex_mul to fire.
